File: chat_hub/infrastructure/repository/message_repository.go
package repository

import (
	entity "chat_hub/core/domain/entities"
	"chat_hub/core/domain/enums"
	base_repo "chat_hub/infrastructure/repository/base"
	"chat_hub/kernel/utils"
	"database/sql"
	"fmt"
	"time"

	"github.com/google/uuid"
)

type MessageRepository struct {
	db   *sql.DB
	base *base_repo.DB
}

func NewMessageRepository(db *sql.DB) *MessageRepository {
	return &MessageRepository{
		db:   db,
		base: base_repo.NewDB(db),
	}
}

var (
	ChatMessageTable = `messages`
)

// func (r *UserMessageRepository) CreateUserMessage(request request_dtos.MessageRequestDTO) (string, error) {
// 	var entity entity.MessageEntity
// 	base_repo.ConvertStruct(request, &entity)
// 	entity.ID = uuid.New().String()
// 	columns, values := base_repo.StructToColumnsAndValues(entity)
// 	id, err := r.base.InsertDB(UserChatMessageTable, columns, values)
// 	if err != nil {
// 		return "", err
// 	}
// 	return id, nil
// }

func (r *MessageRepository) CreateMessage(request entity.MessageEntity) (string, error) {
	request.ID = uuid.New().String()
	columns, values := base_repo.StructToColumnsAndValues(request)
	id, err := r.base.InsertDB(ChatMessageTable, columns, values)
	if err != nil {
		return "", err
	}
	return id, nil
}

// GetRecentChatMessagesByDialogueId retrieves all messages in a dialogue starting
// from the N-th latest user message. It returns every message from that point
// onward ordered by creation time descending.
func (r *MessageRepository) GetRecentChatMessagesByDialogueId(dialogueId string, numberOfMessages int) ([]entity.MessageEntity, error) {
	query := `SELECT user_id, dialogue_id, sender_type, content, metadata
                FROM messages
                WHERE dialogue_id = $1
                AND created_at >= COALESCE((
                        SELECT created_at FROM messages
                        WHERE dialogue_id = $1 AND sender_type = $2
                        ORDER BY created_at DESC
                        OFFSET $3 LIMIT 1
                ), to_timestamp(0))
                ORDER BY created_at DESC`
	rows, err := r.db.Query(query, dialogueId, string(enums.UserMessage), numberOfMessages-1)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var messages []entity.MessageEntity
	for rows.Next() {
		var message entity.MessageEntity
		if err := rows.Scan(&message.UserId, &message.DialogueId,
			&message.SenderType, &message.Content, &message.Metadata); err != nil {
			return nil, err
		}
		messages = append(messages, message)
	}
	return messages, nil
}

func (r *MessageRepository) GetMessagesByDialogueIdWithCursorPagination(dialogueId string, size int, cursor string) ([]entity.MessageEntity, error) {
	size = utils.ValidatePagination(size)

	var (
		rows *sql.Rows
		err  error
	)

	baseQuery := `
		SELECT id, user_id, dialogue_id, sender_type, content, metadata, created_at
		FROM messages
		WHERE dialogue_id = $1
	`

	if cursor == "" {
		query := baseQuery + `
			ORDER BY created_at DESC
			LIMIT $2`
		rows, err = r.db.Query(query, dialogueId, size)
	} else {
		query := baseQuery + `
			AND created_at < $2::timestamp
			ORDER BY created_at DESC
			LIMIT $3`
		rows, err = r.db.Query(query, dialogueId, cursor, size)
	}

	if err != nil {
		return nil, fmt.Errorf("error executing query: %w", err)
	}
	defer rows.Close()

	var messages []entity.MessageEntity
	for rows.Next() {
		var message entity.MessageEntity
		var createdAt time.Time
		if err := rows.Scan(&message.ID, &message.UserId, &message.DialogueId, &message.SenderType,
			&message.Content, &message.Metadata, &createdAt); err != nil {
			return nil, fmt.Errorf("error scanning row: %w", err)
		}
		message.CreatedAt = createdAt
		messages = append(messages, message)
	}

	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("error iterating rows: %w", err)
	}

	return messages, nil
}

File: ai_core/infrastructure/repository/message_repository.py
import datetime
import uuid
from typing import List, Tuple

from core.domain.entities.message import Message
from kernel.database.base import BaseRepository
from kernel.database.postgres import postgres_db


class MessageRepository(BaseRepository[Message]):
    def __init__(self):
        super().__init__(
            table_name="messages",
            model_cls=Message,
            pk="id",
            default_order_by="created_at DESC",
        )

    async def create_message(self, message: Message) -> str:
        """
        Create a new message and return its ID.
        """
        message.id = uuid.uuid4()
        ret = await self.insert(
            message,
            returning=("id", "user_id", "dialogue_id", "user_message_id", "message_type", 
                       "sender_type", "content", "metadata", "created_at", "updated_at"),
            auto_timestamps=True,
        )
        return ret["id"]

    # --- GetRecentChatMessagesByDialogueId ---
    async def get_recent_chat_messages_by_dialogue_id(
        self, dialogue_id: str, number_of_messages: int
    ) -> List[Message]:
        """
        Retrieve recent messages in a dialogue starting from the N-th latest user message.
        """
        # Perform a subquery to get the N-th latest user message
        query = """
            SELECT user_id, dialogue_id, sender_type, content, metadata
            FROM messages
            WHERE dialogue_id = $1
            AND created_at >= COALESCE((
                SELECT created_at
                FROM messages
                WHERE dialogue_id = $1 AND sender_type = 'user'
                ORDER BY created_at DESC
                OFFSET $2 LIMIT 1
            ), to_timestamp(0))
            ORDER BY created_at DESC
            LIMIT $3
        """
        pool = await postgres_db.connect()
        async with pool.acquire() as conn:
            rows = await conn.fetch(query, dialogue_id, number_of_messages - 1, number_of_messages)

        return [self._row_to_model(self.model_cls, row) for row in rows]

    # --- GetMessagesByDialogueIdWithCursorPagination ---
    async def get_messages_by_dialogue_id_with_cursor_pagination(
        self, dialogue_id: str, size: int, cursor: str
    ) -> Tuple[List[Message], bool]:
        """
        Get messages by dialogue with cursor pagination.
        """
        size = max(1, min(size, 100))  # Pagination validation
        query_limit = size + 1  # For checking if more items exist

        # Basic query for message retrieval
        base_query = """
            SELECT id, user_id, dialogue_id, sender_type, message_type, content, metadata, created_at
            FROM messages
            WHERE dialogue_id = $1
        """

        query = base_query
        params = [dialogue_id]

        if cursor:
            query += " AND created_at < $2::timestamp"
            query += " ORDER BY created_at DESC LIMIT $3"
            params.append(datetime.datetime.fromisoformat(cursor))
            params.append(query_limit)
        else:
            query += " ORDER BY created_at DESC LIMIT $2"
            params.append(query_limit)

        pool = await postgres_db.connect()
        async with pool.acquire() as conn:
            rows = await conn.fetch(query, *params)

        # Transform rows into message models
        messages = []
        for row in rows:
            message = self._row_to_model(self.model_cls, row)
            messages.append(message)

        has_more = len(messages) > size
        if has_more:
            messages = messages[:size]

        return messages, has_more

message_repository = MessageRepository() 
